README.md

[![Build Status](https://travis-ci.org/citizenos/citizenos-fe.svg?branch=master)](https://travis-ci.org/citizenos/citizenos-fe)
[![Crowdin](https://d322cqt584bo4o.cloudfront.net/citizenos-fe/localized.svg)](https://crowdin.com/project/citizenos-fe)


# CitizenOS-FE

CitizenOS front-end web application - https://app.citizenos.com.

## Running

### Prerequisites

* Node.JS >= 6.13.1 (https://github.com/mklement0/n-install)

### Installing

* Get the source - `git clone git@github.com:citizenos/citizenos-fe.git`
* Go to the source directory - `cd citizenos-fe`
* **DEV ONLY** Add to dev.citizenos.com to your hosts file - `sudo -- sh -c -e "echo '127.0.0.1 dev.citizenos.com' >> /etc/hosts"`
* **DEV ONLY** When using over HTTPS you need to add `./config/certs/citizenosCARoot.pem` to your trusted CA certificate store or browsers will complain.

### Running

* Set `NODE_ENV` environment variable:
    * Production - `production`
    * Test - `test`
    * Development/local - `development`
* Run the application:
    * Production/test - `node ./bin/www`. Note that Heroku and some other cloud envs pick up the starting point from `Procfile` (https://github.com/citizenos/citizenos-fe/blob/master/Procfile.txt)
    * Development - `npm run dev`
* Open 
    * Production/test - your configured public app url.
    * Development - https://dev.citizenos.com:3001 or http://dev.citizenos.com:3000 in your browser.

**NOTES:**

* By default the app runs against CitizenOS public testing environment (https://test.api.citizenos.com). If you want to run against your local [citizenos-api](https://github.com/citizenos/citizenos-api) instance, modify the url in app.js to https://dev.api.citizenos.com:3003

## Contributing

### Pull requests

* All pull requests to `master` branch
* Live site runs on `prod` branch


CHANGELOG.md
## 2021-11-17 - v. 0.1.5

* Add "Manage participants" view under topic edit view
* Alter "My topics" users and groups to support ordering in querys

## 2021-11-11 - v. 0.1.4

* Dashboard - show extra User contact info in the Topic/Group members list for Topic/Group admins and moderators - https://github.com/citizenos/citizenos-fe/issues/670

## 2021-11-04 - v. 0.1.3

* Group share / join - https://github.com/citizenos/citizenos-fe/issues/325

## 2021-10-28

* Update argument section. All actions are in dropdown, added copy direct link functionality to "direct link" - https://github.com/citizenos/citizenos-fe/issues/361

## 2021-10-25

* In topic view changed "Settings" to "Actions". Altered "Leave topic" confirmation modal to be more descriptive of what that action does

## 2021-10-04

* FIX - Overview menu item overlaps other menu items in mobile view - /my/topics/:topicId - https://github.com/citizenos/citizenos-fe/issues/779

## 2021-09-28 - v. 0.1.2

* Add duplicate topic flow - https://github.com/citizenos/citizenos-fe/issues/416

## 2021-05-31  - v. 0.1.1

* App logout to call `GET /ep_auth_citizenos/logout` to destroy EP session - https://github.com/citizenos/citizenos-fe/issues/676
* Added support for `CITIZENOS_ETHERPAD_BASEURL` env variable to configure Etherpad base URL - https://github.com/citizenos/citizenos-fe/issues/676

## 2021-05-11

* **FE now requires CitizenOS API >= 2.2.0.**
* Localization: GroupMember model renamed to GroupMemberUser - https://github.com/citizenos/citizenos-api/issues/198

## 2020-09-30

* Report a Topic - hide/show moderated topics - https://github.com/citizenos/citizenos-fe/issues/464

## 2020-09-24

* Make entering topic view mode from my topics easier - https://github.com/citizenos/citizenos-fe/issues/237

## 2020-09-23

* 'See who liked your argument' feature not working https://github.com/citizenos/citizenos-fe/issues/572
* No information about voting deadline passed - https://github.com/citizenos/citizenos-fe/issues/502

## 2020-09-22

* Update dependencies and related deprecated code
* Add missing translation keys - https://github.com/citizenos/citizenos-fe/issues/551

## 2020-09-18

* Activity feed shows incorrect numbers in grouped activities - https://github.com/citizenos/citizenos-fe/issues/496
* Opinion festival box for front page- https://github.com/citizenos/citizenos-fe/issues/556

## 2020-09-16

* Etherpad language not always in sync with app UI language - https://github.com/citizenos/citizenos-fe/issues/504
* Sharing private topics displays translations keys as title and description - https://github.com/citizenos/citizenos-fe/issues/543

## 2020-09-15

* Enable account termination by user - https://github.com/citizenos/citizenos-fe/issues/318

## 2020-09-14

* When voting is active, open "Vote" tab by default - https://github.com/citizenos/citizenos-fe/issues/513
* Drop DigiDocService and Proxy using for ID-Card signing - https://github.com/citizenos/citizenos-api/issues/157

## 2020-09-04

* Make search more visible and only when it is relevant - https://github.com/citizenos/citizenos-fe/issues/236
* Search bar should provide feedback if user's input isn't suitable for search - https://github.com/citizenos/citizenos-fe/issues/324

## 2020-08-31

* Hide disabled features (links, buttons) - https://github.com/citizenos/citizenos-fe/issues/240

## 2020-08-19

* Update argument reply votes
* FIX Google scipt outdating for picker - https://github.com/citizenos/citizenos-fe/issues/459
* Change tab name from "View" to "Edit" and back, based on what action user is performing - https://github.com/citizenos/citizenos-fe/issues/302

## 2020-08-17

* Display list of users reactions on arguments - https://github.com/citizenos/citizenos-fe/issues/470

## 2020-07-27

* Enable multiple choice voting - https://github.com/citizenos/citizenos-fe/issues/413

## 2020-07-01

* Update etherpad styles

## 2020-06-17

* Add multiple emails to group

## 2020-05-26

* Topic title and description to meta tags so that social media previews would be generated properly
    * https://github.com/citizenos/citizenos-fe/issues/166
    * https://github.com/citizenos/citizenos-fe/issues/420

## 2020-05-19

* Fix missing comment icon in Etherpad inline comments - https://github.com/citizenos/citizenos-fe/issues/467
* Fix toggling tabs
* Home - hover effect for tiles - slight animation - https://github.com/citizenos/citizenos-fe/issues/305
* Dashboard tabs behaviour fix - https://github.com/citizenos/citizenos-fe/issues/100


## 2020-05-18

* Enable vertical scroll of navigation so that logout would not become inaccessible on small screens - https://github.com/citizenos/citizenos-fe/issues/299
* Delegation - error message when User tries to delegate to a non-topic member - https://github.com/citizenos/citizenos-fe/issues/359

## 2020-05-13

* Fix dashboard topic member list showing inaccessible groups incorrectly - https://github.com/citizenos/citizenos-fe/issues/260
* Fix topic and widget activity feed

## 2020-05-12

* Activity feed direct link to comments - https://github.com/citizenos/citizenos-fe/issues/468
* Activity Feed doesn't show new activities' badge upon login - https://github.com/citizenos/citizenos-fe/issues/472
* Activity Feed stays visible after log-out - https://github.com/citizenos/citizenos-fe/issues/473

## 2020-05-11

* Update activity feed grouping
* Load more activities if list is too short after grouping
* Fix topic invite mobile view

## 2020-05-07

* Topic inivite to separate modal, add pagination, can copy-paste multiple e-mails - https://github.com/citizenos/citizenos-fe/issues/235

## 2020-04-28

* Add pagination funcitonality for topic members


## 2020-04-27

* Group similar Activity Feed items together - https://github.com/citizenos/citizenos-fe/issues/256
* FIX delete topic alignment in Safari - https://github.com/citizenos/citizenos-fe/issues/257

## 2020-04-22


index.js
'use strict';

var config = require('config');
var express = require('express');
var app = express();
var https = require('https');
var path = require('path');
var http = require('http');
var fs = require('fs');
var _ = require('lodash');
const { expressCspHeader, INLINE, NONE, SELF } = require('express-csp-header');


var prerender = require('prerender-node');

var configFe = config.util.loadFileConfigs('./public/config');

var pathSettings = path.resolve('./public/settings.js');
try {
    var settingsFileTxt = '(function (window) { window.__config = window.__config || {};';
    _(configFe).forEach(function (value, key) {
        settingsFileTxt += ' window.__config.' + key + ' = ' + JSON.stringify(value) + ';';
    });
    settingsFileTxt += '}(this));';
    fs.writeFileSync(pathSettings, settingsFileTxt);
} catch (err) {
    console.log('Settings.json write FAILED to ' + pathSettings, err);
    process.exit(1);
}
//TODO: list whitelisted urls to github with description
var cspConfig = config.csp;
var cspOptions = _.cloneDeep(cspConfig);
if (cspConfig) {
    if (cspConfig.directives) {
        if (typeof cspConfig.directives === 'string') {
            cspConfig.directives = JSON.parse(cspConfig.directives);
            cspOptions.directives = {};
        }
        Object.keys(cspConfig.directives).forEach(function(key, index) {
            cspConfig.directives[key].forEach(function (value, k) {
                if (k === 0) {
                    cspOptions.directives[key] = [];
                }
                if (value === 'none') {
                    cspOptions.directives[key].push(NONE);
                } else if (value === 'self') {
                    cspOptions.directives[key].push(SELF);
                } else if (value === 'inline') {
                    cspOptions.directives[key].push(INLINE);
                } else {
                    cspOptions.directives[key].push(value);
                }
            });
        });
    }

    app.use(expressCspHeader(cspOptions));
}


app.use(prerender.set('prerenderToken', 'CrrAflHAEiF44KMFkrs7'));

app.use(express.static(__dirname + '/public'));

const browserDetect = (req, res, next) => {
    const ua = req.headers['user-agent'];

    const msie = ua.indexOf('MSIE ');
    if (msie > 0) {
        // IE 10 or older => return version number
        var IEversion =  parseInt(ua.substring(msie + 5, ua.indexOf('.', msie)), 10);
        console.log('IEversion', IEversion);
        res.sendFile(__dirname + '/public/views/unknown_device.html');

        return res.set('Permissions-Policy', 'interest-cohort=()'); // Opt-out of Google FLoC
    }

    const trident = ua.indexOf('Trident/');
    if (trident > 0) {
        // IE 11 => return version number
        var rv = ua.indexOf('rv:');
        var version =  parseInt(ua.substring(rv + 3, ua.indexOf('.', rv)), 10);
        console.log('IE 11');
        res.sendFile(__dirname + '/public/views/unknown_device.html');

        return res.set('Permissions-Policy', 'interest-cohort=()'); // Opt-out of Google FLoC
    }
    // other browser

    next();
}

app.use(browserDetect);
app.get('/*', browserDetect, function (req, res) {
    res.sendFile(__dirname + '/public/index.html');
    res.set('Permissions-Policy', 'interest-cohort=()'); // Opt-out of Google FLoC
});

var host = process.env.HOST || null;

var portHttp = process.env.PORT || 3000;
http.createServer(app).listen(portHttp, host, function (err, res) {
    if (err) {
        console.log('Failed to start HTTP server on port' + portHttp, err);
        return;
    }
    console.log('HTTP server listening on port ' + portHttp);
});


if (app.get('env') === 'development') {
    var portHttps = process.env.PORT_SSL || 3001;
    var options = {
        key: fs.readFileSync('./config/certs/dev.citizenos.com.key'),
        cert: fs.readFileSync('./config/certs/dev.citizenos.com.crt')
    };

    https.createServer(options, app).listen(portHttps, host, function (err, res) {
        if (err) {
            console.log('Failed to start HTTPS server on port' + portHttps, err);
            return;
        }
        console.log('HTTPS server listening on port ' + portHttps);
    });
}
